pallets/pallet-auction/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+			let id = ensure_signed(origin)?;
+
+			// Ensure role is customer.
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// check the NftIdState is Free
+			T::NftManager::ensure_nft_state(nft_id, NftState::Free)?;
+
+			// Check the Nft is belong to the origin
+			T::NftManager::ensure_nft_is_valid(&id, nft_id)?;
+
+			// Set the Auctions storage
+			let auction_id = Self::next_auction_id();
+			let expiry_block =
+				T::AuctionLength::get() + frame_system::Pallet::<T>::current_block_number();
+
+			Auctions::<T>::insert(
+				auction_id,
+				AuctionData {
+					nft_id,
+					owner: id.clone(),
+					start,
+					reserve,
+					buy_now,
+					expiry_block,
+					current_bid: None,
+				},
+			);
+
+			// Append the auction id to the AuctionsExpiryBlock storage.
+			AuctionsExpiryBlock::<T>::append(expiry_block, auction_id);
+
+			// Change Nft state to auction.
+			T::NftManager::change_nft_state(nft_id, NftState::Auction(auction_id))?;
+
+			// Pay fee.
+			T::Bank::transfer(&id, &T::Bank::treasury()?, T::AuctionStartFee::get())?;
+
+			Self::deposit_event(Event::<T>::AuctionCreated { who: id, auction_id });
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::bid())]
+		pub fn bid(
+			origin: OriginFor<T>,
+			auction_id: AuctionId,
+			price: T::Balance,
+		) -> DispatchResult {
+			// Get the account id
+			let id = ensure_signed(origin)?;
+
+			// Ensure role is customer.
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// update storage auctions
+			Auctions::<T>::try_mutate(auction_id, |maybe_auction_data| {
+				if let Some(auction_data) = maybe_auction_data.as_mut() {
+					let current_price = sp_std::cmp::max(
+						if let Some((last_bidder, last_price)) = &auction_data.current_bid {
+							// Refund the previous bidder's bid
+							T::Bank::transfer(
+								&T::BidsPoolAccount::get(),
+								last_bidder,
+								*last_price,
+							)?;
+							Ok::<T::Balance, DispatchError>(*last_price)
+						} else {
+							Ok(Default::default())
+						}?,
+						auction_data.start.unwrap_or_default(),
+					);
+
+					// Ensure bid is valid.
+					ensure!(
+						price >= current_price + T::MinimumIncrease::get(),
+						Error::<T>::BidPriceTooLow
+					);
+
+					// Transfer bid to Bids Pool's account.
+					T::Bank::transfer(&id, &T::BidsPoolAccount::get(), price)?;
+
+					// Update current bid's storage.
+					auction_data.current_bid = Some((id.clone(), price));
+
+					// Calculate how many blocks the auction will over, if it shorter than specific
+					// length, then extend to specific length.
+					if T::ExtendedLength::get() >
+						auction_data.expiry_block -
+							frame_system::Pallet::<T>::current_block_number()
+					{
+						// Remove the old auction id
+						AuctionsExpiryBlock::<T>::mutate(auction_data.expiry_block, |auctions| {
+							auctions.retain(|id| *id != auction_id);
+						});
+
+						let new_expiry = frame_system::Pallet::<T>::current_block_number()
```

Comment:
You can do this calculation first, then use this number to compare. This calculation
new_expiry = (current + extend) > expiry
 is the same as before
extend > expiry - current